cmd/snap-gpio-helper/common.go

```diff
@@ -0,0 +1,235 @@
+	"github.com/snapcore/snapd/strutil"
+)
+
+const (
+	aggregatorLockPath         = "/sys/bus/platform/drivers/gpio-aggregator"
+	aggregatorNewDevicePath    = "/sys/bus/platform/drivers/gpio-aggregator/new_device"
+	aggregatorDeleteDevicePath = "/sys/bus/platform/drivers/gpio-aggregator/delete_device"
+	ephermalUdevRulesDir       = "/run/udev/rules.d"
+)
+
+var lockAggregator = func() (unlocker func(), err error) {
+	flock, err := osutil.OpenExistingLockForReading(filepath.Join(dirs.GlobalRootDir, aggregatorLockPath))
+	if err != nil {
+		return nil, err
+	}
+	if err := flock.Lock(); err != nil {
+		return nil, err
+	}
+	return func() {
+		flock.Close()
+	}, nil
+}
+
+var aggregatorCreationTimeout = 120 * time.Second
+var deviceGetGpioChardevChipInfo = device.GetGpioChardevChipInfo
+
+func addAggregatedChip(sourceChip *device.GPIOChardev, commaSeparatedLines string) (chip *device.GPIOChardev, err error) {
+	// synchronize gpio helpers' access to the aggregator interface
+	unlocker, err := lockAggregator()
+	if err != nil {
+		return nil, err
+	}
+	defer unlocker()
+
+	f, err := os.OpenFile(filepath.Join(dirs.GlobalRootDir, aggregatorNewDevicePath), os.O_WRONLY, 0)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	watcher, err := inotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	err = watcher.AddWatch(dirs.DevDir, inotify.InCreate)
+	if err != nil {
+		return nil, err
+	}
+
+	// <label> <lines>
+	_, err = fmt.Fprintf(f, "%s %s", sourceChip.Label, commaSeparatedLines)
+	if err != nil {
+		return nil, err
+	}
+
+	ctxWithTimeout, cancel := context.WithTimeout(context.Background(), aggregatorCreationTimeout)
+	defer cancel()
+	for {
+		select {
+		case event := <-watcher.Event:
+			path := event.Name
+			// check prefix /dev/gpiochipX
+			if strings.HasPrefix(path, filepath.Join(dirs.DevDir, "gpiochip")) {
+				return deviceGetGpioChardevChipInfo(path)
+			}
+		case <-ctxWithTimeout.Done():
+			return nil, fmt.Errorf("max timeout exceeded")
+		}
+	}
+}
+
+func aggregatedChipUdevRulePath(instanceName, slotName string) string {
+	fname := fmt.Sprintf("69-snap.%s.interface.gpio-chardev-%s.rules", instanceName, slotName)
+	return filepath.Join(filepath.Join(dirs.GlobalRootDir, ephermalUdevRulesDir), fname)
+}
+
+func addEphermalUdevTaggingRule(chip *device.GPIOChardev, instanceName, slotName string) error {
+	if err := os.MkdirAll(filepath.Join(dirs.GlobalRootDir, ephermalUdevRulesDir), 0755); err != nil {
+		return err
+	}
+
+	tag := fmt.Sprintf("snap_%s_interface_gpio_chardev_%s", instanceName, slotName)
+	rule := fmt.Sprintf("SUBSYSTEM==\"gpio\", KERNEL==\"%s\", TAG+=\"%s\"\n", chip.Name, tag)
+
+	path := aggregatedChipUdevRulePath(instanceName, slotName)
+	if err := os.WriteFile(path, []byte(rule), 0644); err != nil {
+		return err
+	}
+
+	// make sure the rule we just dropped is loaded as sometimes it doesn't get
+	// picked up right away
+	output, err := exec.Command("udevadm", "control", "--reload-rules").CombinedOutput()
```

Comment:
This should be `exec.CommandContext` and the context should be given as argument. This way the entire chain of actions can be aborted when something interrupts the `snap-gpio-helper` program.